src/os/bluestore/BlueStore.h

```diff
@@ -519,17 +480,49 @@ class BlueStore : public ObjectStore,
     void dump(BufferCacheShard* cache, ceph::Formatter *f) const {
       std::lock_guard l(cache->lock);
       f->open_array_section("buffers");
-      for (auto& i : buffer_map) {
+      for (auto& b : buffer_map) {
 	f->open_object_section("buffer");
-	ceph_assert(i.first == i.second.offset);
-	i.second.dump(f);
+	b.dump(f);
 	f->close_section();
       }
       f->close_section();
     }
     friend std::ostream& operator<<(std::ostream& out, const BufferSpace& bc);
   };
 
+  class Writings {
+    struct WriteEntry {
+      Onode* onode;
+      uint32_t offset;
+      uint32_t length;
+      WriteEntry(Onode* _o, Buffer* b)
+        : onode(_o), offset(b->offset), length(b->length) {}
+    };
+
+    using write_list_t = mempool::bluestore_writing::list<WriteEntry>;
+
+    ceph::mutex lock = ceph::make_mutex("BlueStore::Writings::lock");
+
+    // TransContext -> list of <onode, offset, length>
+    mempool::bluestore_writing::map<void*, write_list_t> txc_to_buf;
+
+  public:
+    ~Writings() {
+      // sanity
+      ceph_assert(txc_to_buf.empty());
+    }
+    void add_writing(Onode* o, Buffer* b) {
+      std::lock_guard l(lock);
+      txc_to_buf[b->txc].emplace_back(o, b);
+    }
+    void finish_writing(TransContext* txc);
+
+    bool empty() {
+      std::lock_guard l(lock);
+      return txc_to_buf.empty();
+    }
+  } writings;
+
   struct SharedBlobSet;
```

Comment:
I do not get what we get from centralized `Writings`.

We need:
1) Prevent buffers from disposal until data is stable on disk; that is, until  `_txc_finish`.
2) When we clone writing buffers we must make sure target BufferSpace is also notified.
    Specifically `_dup_writing` must do something to secure it.
My assumptions/preferences:
3) Writing buffers are few per onode.
4) Data cache is usually small, there are few buffers per `BufferSpace` too.
5) Clone is done within single collection.
6) Smaller c++ footprint is always better.

`BufferSpace::writing` was sorted by seq,
now `Writings::txc_to_buf` are sorted by ctx.
I will be a larger container (slower,worse).

`BufferSpace::writting` was replaced with `BufferSpace::onode`. (less space, better)

`BufferSpace::writting` introduces new lock. I assume that if it were to be duplicated to every Collection,
then the need for a lock would vanish.

@ifed01 Is this work a preparation for some future steps?

Reply:
Hi @aclamk. 
My general intention was to start improving the design of this part of the code. Current monolithic structure looks unclear, error-prone, untestable and hence hard to maintain. 
With Pere's PR it becomes even worse, e.g. see new deferred_seq_dependencies list: 
specific write handling (deferred one) needs additional stuff at Collection level to track onodes which monitor pending writes in order to ensure consistent data caching. Brrr...
And this has to be properly updated/copied on onode cloning and collection splitting.
Additionally we tracked pending writes at BufferSpace... 

IMO that's a terrible design. And I believe it's worth improving even if we get some overhead (which I believe is pretty neglectable) after the redesign.
So yes I'm introducing one more lock - which is required indeed if we access centralized data collection. But the duration this lock is acquired should be pretty short - just an insertion into a map and/or a list. And these collections Which shouldn't be very large as they're tracking writings entries only. 
BTW we can consider changing txc_to_seq map into another intrusive list - insertion/removal overhead would be tiny then.

And asking your last question - I'm not planning to proceed with further redesign of the buffer caching in short-term though..

Reply:
well, the above discussion inspired me for another rework.. Global Writings has been removed and TransContext keeps tracking Onode observers for pending writings. Still needs a lock but it's per-TransContext now. But the real benefit is in no need for global txc->observer mapping any more. Hence no need for expensive lookup and insertion/removal.
TransContext just keeps an unsorted list of observers..
how do you like this, @aclamk ?

Reply:
@ifed01 The solution is good, I like it.
I reviewed it and I think its good to go.